### PR TITLE
Add custom http headers

### LINKED ISSUE #8656: Feature Request: Domain Fronting for meterpreter/mettle
It would be really amazing to allow meterpreter, when using the http or https transports, to take advantage of domain fronting. This SHOULD only require the addition of a custom host header to the outbound requests. I am not familiar enough with the meterpreter internals to be able to say if it affects other components of meterpreter, or would cause a problem, though.

Just wishful thinking :D

### PR DESCRIPTION
**Pull request by OJ**
With "domain fronting" becoming all the rage it was high time we added some kind of support to the Meterpreter payloads that would allow for these shenanigans to work in MSF. This PR is the first attempt at such a thing, while aiming to add support for other things including `Referer` and `Cookie`. The latter field may be helpful for generated payloads that need to get out through something like Bluecoat.

PR marked as delayed while we work through a few TODOs and I get help from @busterb, @timwr and @zeroSteiner to polish things up.

The associated Payloads PR is here: https://github.com/rapid7/metasploit-payloads/pull/236
The associated Mettle PR is here: <coming shortly> (Will ask Brent for help on this one)

# Generic headers

Obviously the need to set generic headers is helpful. Unfortunately, opening this door to people can cause issues because not all headers can be set in the way that we'd expect, and this varies from API to API and platform to platform. The `Host` header in particular can be quirky (as can be seen with the `Python` implementation).

As a result, I added support for three new settings which map directly to headers:

* `HttpHeaderHost`
* `HttpHeaderReferer`
* `HttpHeaderCookie`

These can be handled on a per-implementation basis and put in the correct spots. If we feel the need to add more over time, then we can.

# Domain Fronting

To support domain fronting, we really only needed to give control over the `Host` header. Hence to make this work in a payload:

* Set the `LHOST` and `LPORT` parameters to what you'd expect the payload to connect to.
* Set the `HttpHeaderHost` to the fronted domain.

Off you go. From here the requests should include the custom Host header. If you're using the likes of Amazon with Cloudfront, then things should just work as you'd expect.

# Verification

- [x] Create payloads and handlers for all the `http` payloads using the new options. For my testing I pointed to my local listener (both staged and stageless).
- [x] Fire up Wireshark and make sure the filter is set up so that HTTP traffic is captured on the give port(s) that you have listeners on
- [x] Launch the payloads, and view the content in wireshark.
- [x] Verify that the connections all reach out to the address specified in `LHOST`.
- [x] Verify that the connections contain the configured host header properties.
- [x] Make sure that all the sessions work.

# TODO

- [x] Make sure this works with the multi payload type.
- [x] Get the Mettle code going (cc @busterb).
- [x] Get the java payloads going (cc @timwr).

Thanks!

Fixes #8656

### CODE REVIEW

**Comment by Chiggins** [comment]
Looking slick, good job OJ. If you need anyone to help testing this out, let me know, I've already got infrastructure set up for domain fronting.

**Comment by OJ** [comment]
Thanks @Chiggins, got a few more kinks to work out, but I'll take you up on that!

File: lib/msf/core/payload/transport_config.rb [hunk 1]
```diff
@@ -80,6 +81,13 @@ def transport_config_reverse_named_pipe(opts={})
 
 private
 
+  def get_custom_headers(ds)
+    headers = ""
+    headers << "Host: #{ds['HttpHost']}\r\n" if ds['HttpHost']
```
Comment (timwr) [inline lib/msf/core/payload/transport_config.rb]: HttpHeaderHost ?

File: lib/msf/core/payload/windows/reverse_http.rb [hunk 2]
```diff
@@ -36,7 +36,10 @@ def initialize(*args)
         OptPort.new('PayloadProxyPort', [false, 'An optional proxy server port']),
         OptString.new('PayloadProxyUser', [false, 'An optional proxy server username']),
         OptString.new('PayloadProxyPass', [false, 'An optional proxy server password']),
-        OptEnum.new('PayloadProxyType', [false, 'The type of HTTP proxy (HTTP or SOCKS)', 'HTTP', ['HTTP', 'SOCKS']])
+        OptEnum.new('PayloadProxyType', [false, 'The type of HTTP proxy (HTTP or SOCKS)', 'HTTP', ['HTTP', 'SOCKS']]),
+        OptString.new('HttpHeaderHost', [false, 'An optional value to use for the Host HTTP header']),
```
Comment (timwr) [inline lib/msf/core/payload/windows/reverse_http.rb]: Might be worth moving these options to a mixin that can be included to lib/msf/core/payload/windows/reverse_http.rb, lib/msf/core/payload/android/reverse_http.rb
lib/msf/core/payload/python/reverse_http.rb

**Review by timwr — COMMENTED** [review]
@OJ I'll PR you the Android/Java fixes asap

**Comment by timwr** [comment]
I've PRed some fixes for Android/Java here:
https://github.com/OJ/metasploit-framework/pull/20
https://github.com/OJ/metasploit-payloads/pull/8
I had to add OptString.new('HttpHeaderHost', etc to lib/msf/core/payload/(android|java)/reverse_http.rb to make it work but I don't think that's the way to go. We should consider moving those options somewhere to avoid duplication.
Also perhaps we should alias MeterpreterUserAgent to HttpHeaderUserAgent.
Thoughts?

**Comment by OJ** [comment]
Thanks @timwr I've merged both of the PRs. I do think that we need to look to reduce the duplication of code, but it's hard to know where best to apply it. Plus if we're going to do it here, we should look to do it in a number of other spots where duplication is rampant.

**Comment by busterb** [comment]
Maybe we could just define some helper methods in the base class like 'register_http_options' that are called by each meterpreter that implements them.

File: lib/msf/core/payload/python/reverse_http.rb [hunk 3]
```diff
@@ -24,11 +27,14 @@ def initialize(info = {})
...
-      host:       ds['LHOST'] || '127.127.127.127',
-      port:       ds['LPORT'],
-      proxy_host: ds['PayloadProxyHost'],
-      proxy_port: ds['PayloadProxyPort'],
-      user_agent: ds['MeterpreterUserAgent']
+      host:           ds['LHOST'] || '127.127.127.127',
+      port:           ds['LPORT'],
+      proxy_host:     ds['PayloadProxyHost'],
+      proxy_port:     ds['PayloadProxyPort'],
+      user_agent:     ds['MeterpreterUserAgent'],
```
Comment (busterb) [inline lib/msf/core/payload/python/reverse_http.rb]: Sorry if this was a pre-existing issue, is it OK if we rename 'MeterpreterUserAgent' to 'HttpUserAgent' for consistency? You can modify where it's registered to have 'MeterpreterUserAgent' as an alias. Plus, this field gets used by stagers, so not necessarily a Meterpreter-only field anyway.

**Comment by busterb** [comment]
Does the HTTP Cookies option imply that the payload should implement full-fledged cookie tracking? Or that we should fill an actual cookie-jar with assorted cookies?

Reason why I ask is I can either just add the header, or do the real deal: https://curl.haxx.se/libcurl/c/CURLOPT_COOKIE.html and wondering which is better for this purpose.

**Comment by busterb** [comment]
The mettle-specific PR is above, note that the --cookie option is a little more complex than I think we are exposing here. I think I'll just make it assume sane defaults rather than requiring the user to set a half-dozen knobs per cookie.

**Comment by busterb** [comment]
BTW, my other changes to this PR are at https://github.com/OJ/metasploit-framework/pull/21

**Comment by OJ** [comment]
This should be good to go. I'll look into the conflicts (which really don't make sense!)

**Comment by OJ** [comment]
Selfish Bump! Would someone mind kicking off a new Travis build please?

**Comment by pbarry-r7** [comment]
Hmm, looks like there's a [mismatch on payload sizes](https://travis-ci.org/rapid7/metasploit-framework/jobs/283002266) causing the Travis failures...

**Review by timwr — APPROVED** [review]
Works well on Android/OSX

**Comment by busterb** [comment]
Rebased on master since the options handling code is now there. If this passes specs, will continue testing, adding mettle support, and ship.

**Comment by busterb** [comment]
mettle support is now pushed up now!

**Comment by busterb** [comment]
Landed, thanks all!

**Comment by busterb** [comment]
## Release Notes

This adds initial support for modifying HTTP headers with Meterpreter for use in domain fronting and other evasion applications.

**Comment by c0rpse** [comment]
God job.

**Pull request merged**